File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def __init__(self):
        self.result = []

    def dfs(self, curr, candidates, target, currSum, currList):
        if currSum > target:
            return
        if currSum == target:
            res = currList.copy()
            self.result.append(res)
            return
        for i in range(curr+1, len(candidates)):
            if i > curr + 1 and candidates[i-1] == candidates[i]:
                continue
            currList.append(candidates[i])
            self.dfs(i, candidates, target, currSum + candidates[i], currList)
            currList.pop()
        return

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        for i in range(len(candidates)):
            if i > 0 and candidates[i-1] == candidates[i]:
                continue 
            self.dfs(i, candidates, target, candidates[i], [candidates[i]])
        return self.result

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private List<List<Integer>> result;

    private void backTrack(int[] candidates, int target, int curr, int currSum, LinkedList<Integer> currRes) {
        if (currSum > target) {
            return;
        }
        if (currSum == target) {
            this.result.add(new LinkedList<>(currRes));
            return;
        }
        for (int i = curr; i < candidates.length; i++) {
            if (i > curr && candidates[i] == candidates[i-1]) {
                continue;
            }
            currRes.addLast(candidates[i]);
            this.backTrack(candidates, target, i+1, currSum + candidates[i], currRes);
            currRes.removeLast();
        }
    }


    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        this.result = new LinkedList<>();
        Arrays.sort(candidates);
        this.backTrack(candidates, target, 0, 0, new LinkedList<>());
        return this.result;
    }
}